Adds support for managing edge dictionaries

benzvan opened the pull request:
This is addressing issue #153 

~I'm hoping someone can take a look at this. The test will fail because the help function isn't printing what is expected. I still don't quite understand the code well enough to debug this. The `edgedictionary/describe.go` code is nearly identical to the `healthcheck/describe.go` code and the `root.go` files are also basically the same. But somehow the output of the help is different between the two.~

benzvan commented:
Never mind. Found my problem.

benzvan commented:
I have plans to add functions for `create`, `update`, and `delete` as well. I can either add them to this PR or I can wait until this is merged and add them individually, whichever makes sense for the flow of this project.

benzvan commented:
After some consideration, I decided `describe` wasn't complete enough for a version update and continued working. This is still work in progress and only has describe and create at this point.

benzvan commented:
I believe this to be a releasable update with minimum viable features for dictionaries and dictionary items. I apologize for the size of this change, but each commit is relatively self-contained for easier reviewing. 

My one concern is that if one wanted to create a new dictionary item one would need to do the following:
* `fastly service search --name <servicename>`
* get the service ID and service version from the above command
* `fastly service-version clone ...`
* get the service version from the above command
* `fastly dictionary create ...`
* get the dictionary ID from the above command
* `fastly service-version activate ...`
* `fastly dictionaryitem create ...`

Similarly updating an existing dictionary item will require describing the dictionary to get the dictionary id first.

I would propose SERVICE-ID be an optional flag provided SERVICE-NAME is present and can be used to query fastly for the service ID. Same with DICTIONARY-ID and DICTIONARY-NAME.

I would also propose a flag like `clone-service=true` for `dictionary create` that would combine the first 6 steps listed above so that dictionaries can be created and applied to a service in a single step.

benzvan commented:
Just checking if anyone has had a chance to look at this. @phamann ?

phamann reviewed (changes requested):
@benzvan apologies for the delay here, I appreciate both your effort for the contribution and the poke, so thank you.

Overall, this is looking good. A couple of small points around UX and consistency with the rest of the project. But nothing major. 



imilarly updating an existing dictionary item will require describing the dictionary to get the dictionary id first.

> I would propose SERVICE-ID be an optional flag provided SERVICE-NAME is present and can be used to query fastly for the service ID. Same with DICTIONARY-ID and DICTIONARY-NAME.

This is a good feature request. If you don't want to add the functionality for dictionaries in this PR please feel free to raise this as an issue on the repo and we'll try and get it added to the roadmap.

> I would also propose a flag like clone-service=true for dictionary create that would combine the first 6 steps listed above so that dictionaries can be created and applied to a service in a single step.

Again this is a great feature request, so fell free to add an issue for tracking.

benzvan commented:
Thanks for the feedback. I'll get to work on that and I'll add an issue for the feature enhancement.

benzvan commented on pkg/edgedictionaryitem/create.go:
```diff
@@ -0,0 +1,51 @@
+	c.CmdClause = parent.Command("create", "Create a new item on a Fastly edge dictionary")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("dictionary-id", "Dictionary ID").Required().StringVar(&c.Input.Dictionary)
+	c.CmdClause.Flag("itemkey", "Dictionary item key").Required().StringVar(&c.Input.ItemKey)
+	c.CmdClause.Flag("itemvalue", "Dictionary item value").Required().StringVar(&c.Input.ItemValue)
```
I agree. I thought about this as a question of conformity with the API versus usability of the CLI and erred on the side of conformity.

benzvan commented on pkg/edgedictionary/describe.go:
```diff
@@ -0,0 +1,51 @@
+		return err
+	}
+
+	fmt.Fprintf(out, "Service ID: %s\n", dictionary.ServiceID)
+	fmt.Fprintf(out, "Version: %d\n", dictionary.Version)
+	text.PrintDictionary(out, "", dictionary)
```
Is it possible this comment is on the wrong file? I'm seeing the verbose flag  being used frequently in the `list` functions but never in the `describe` functions. the `PrintDictionary` function already prints `id`, `write_only`, and `created_at`, but I can definitely see the utility of adding `item_count` and `digest` to the output.

phamann replied:
Apologies for the confusion. I hadn't read the implementation of `text.PrintDictionary` before writing the original comment. However, yes I do think that a call to `gofastly.GetDictionaryInfo()` here would be good to get the `last_updated`, `item_count and `digest` fields. We could pass a pointer to the returned struct to PrintDictionary and conditionally print the fields if it exists.

benzvan commented:
I've addressed your comments with new commits and one comment.

I'm willing to add both of the features I mention above within this PR but I'm having trouble figuring out if kingpin supports either/or required flags or if that's something I'll need to add to each file.

benzvan commented:
I'm going to pause for a moment (assuming tests pass) to allow you to look at my latest two commits. I'd like to make sure I'm matching your expectations from a code structure standpoint. I created a new `service.go` file that contains the `GetIDFromName()` function that is used by both the `service describe` and `dictionary create` commands.

benzvan commented:
@phamann just poking you for input again. I'm not sure I'll have time to do more on this for another week or so.

phamann commented:
Hey @benzvan apologies for the delay on this PR. We were focused on getting the AssemblyScript support out first #160. 

With regards two your last two commits, if not too much pain for you, I'd prefer it if you can move the `--service-name` work to its own PR, as its an important feature that warrants being in the change log, which is generated on a per PR basis as its automated. It will also allow us to review it in isolation. Hope thats ok? I should have said this earlier, apologies.

Once thats out, I think this PR should be good to go. 😄

benzvan commented:
Not a problem. I'll pull those commits and start a new branch.

On Thu, Oct 29, 2020, 09:16 Patrick Hamann <notifications@github.com> wrote:

> Hey @benzvan <https://github.com/benzvan> apologies for the delay on this
> PR. We were focused on getting the AssemblyScript support out first #160
> <https://github.com/fastly/cli/pull/160>.
>
> With regards two your last two commits, if not too much pain for you, I'd
> prefer it if you can move the --service-name work to its own PR, as its
> an important feature that warrants being in the change log, which is
> generated on a per PR basis as its automated. It will also allow us to
> review it in isolation. Hope thats ok? I should have said this earlier,
> apologies.
>
> Once thats out, I think this PR should be good to go. 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fastly/cli/pull/159#issuecomment-718782265>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWP6RYVTIEZBNA7HFGONPTSNF2MTANCNFSM4R4VLOQA>
> .
>

benzvan commented:
OK. I force-pushed to my branch without the last two commits and it should be ready to merge.

benzvan commented:
Oh...except I need to pull from master...1 sec.

benzvan commented:
ok. I rebased on master and re-pushed. There's still one comment I replied to that may need addressing.

phamann commented:
Apologies for the delay as always @benzvan. I've now replied to that final comment thread: https://github.com/fastly/cli/pull/159/files#r520448224

Then we'll be good to merge.

benzvan commented:
I thought for a minute about how I would like a verbose describe to look if I was using the command line and added a call to get the dictionary items as well. Pending tests passing, I believe this is ready, @phamann .

phamann reviewed (approved):
This looks good to me now. 🎉 

Thank you SO MUCH for your contribution @benzvan and for your patience and understanding throughout this process. Apologies it's taken so long. Part of the reason has been we're transitioning the CLI project over to a new internal team and owner, @Integralist, which will hopefully mean there won't be as much delay in the future.

Related: I want him to have one final look over the PR and then we will merge and release this week for you.

benzvan commented:
No problem! I understand there's a ton of stuff going on right now so everything is slow and unpredictable. We've just been using a clone of my branch for this functionality for now and will bake it into a container for our CI automation once it's released.

Integralist commented on pkg/api/interface.go:
```diff
@@ -56,6 +56,18 @@ type Interface interface {
+	ListDictionaryItems(*fastly.ListDictionaryItemsInput) ([]*fastly.DictionaryItem, error)
+	GetDictionaryItem(*fastly.GetDictionaryItemInput) (*fastly.DictionaryItem, error)
+	CreateDictionaryItem(*fastly.CreateDictionaryItemInput) (*fastly.DictionaryItem, error)
+	UpdateDictionaryItem(*fastly.UpdateDictionaryItemInput) (*fastly.DictionaryItem, error)
```
For the 'Dictionary item' section can you also add `UpdateDictionaryItems` using the Go-Fastly function [`BatchModifyDictionaryItems `](https://github.com/fastly/go-fastly/blob/master/fastly/dictionary_item.go#L205) which maps to the APIs [bulk update](https://developer.fastly.com/reference/api/dictionaries/dictionary-item/#bulk-update-dictionary-item).

benzvan replied:
Do you think it should be it's own command?
```
  dictionaryitems update --dictionary-id=DICTIONARY-ID --file=FILE
```

 or a variation of the existing one?
```
  dictionaryitem update --dictionary-id=DICTIONARY-ID --file=FILE
```

benzvan replied:
Semantically, the plural makes sense but the API section is still `dictionary item` even if the endpoint is `/service/service_id/dictionary/dictionary_id/items`. My vote is that it would be excessive to create a new top-level command.

phamann replied:
I agree that it would be excessive to create a new top-level command for bulk updates and wouldn't follow our API. And like the idea of a `--file` flag to pass a json object 👍

Integralist replied:
Good shout. I'm happy with this suggestion 👍🏻

Integralist commented on pkg/text/dictionary.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+// PrintDictionary pretty prints a fastly.Dictionary structure in verbose
+// format to a given io.Writer. Consumers can provider an prefix string which
```
```suggestion
// format to a given io.Writer. Consumers can provider a prefix string which
```

benzvan replied:
Just realized `provider` should be `provide` which makes me think this exists widely throughout the code because I likely would have copied this from elsewhere.

Integralist commented on pkg/text/dictionaryitem.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+// PrintDictionaryItem pretty prints a fastly.DictionaryInfo structure in verbose
+// format to a given io.Writer. Consumers can provider an prefix string which
```
```suggestion
// format to a given io.Writer. Consumers can provider a prefix string which
```

Integralist commented on pkg/edgedictionary/edgedictionary_test.go:
```diff
@@ -0,0 +1,279 @@
+`) + "\n"
+
+var describeDictionaryOutputVerbose = strings.TrimSpace(`
+Fastly API token not provided
```
Probably me just not understanding the current usage of the tool, but would we expect an API request to be successful without an API token provided? e.g. as far as I'm aware the relevant API endpoint still requires authentication.

benzvan replied:
This is an existing pattern in the other test classes. I assume the verbose output is different because the api token is not present and is not needed for the purpose of the test.

Integralist replied:
OK cool. Thanks for clarifying 👍🏻

Integralist commented on pkg/edgedictionaryitem/list.go:
```diff
@@ -0,0 +1,54 @@
+		return err
+	}
+
+	fmt.Fprintf(out, "Service ID: %s\n", c.Input.Service)
```
This (and other instances) should be replaced with the `text` helper package equivalent I believe. 

/cc @phamann

benzvan replied:
Good call. I was following an existing pattern here, but it never hurts to have more code correct when the refactor happens later.

Integralist replied:
Awesome, thank you!

benzvan replied:
I fixed this in this PR. I started to fix it in other places but there is a more fundamental change required in several the 900+ uses that needs to be addressed in that text.Output trims whitespace and several of the fmt.Fprintf instances have preceding tabs for pretty printing.

Integralist replied:
@benzvan no problem, if it's a long standing inconsistency then I'll try to address it at some later point in time 🙂  as long as new code added is avoiding the use of fmt in favour of a text package functional equivalent, then that's fine 👍🏻

Integralist commented on pkg/edgedictionaryitem/update.go:
```diff
@@ -0,0 +1,51 @@
+		return err
+	}
+
+	fmt.Fprintf(out, "Service ID: %s\n", c.Input.Service)
```
As mentioned earlier I've just noticed the use of `fmt` output functions which I believe should be replaced with `text` helper package equivalents.

Integralist reviewed (commented):
This is some great work @benzvan. Thank you! 

I have left some comments. 
Please do reach out if you have any questions.

benzvan commented on pkg/api/interface.go:
```diff
@@ -56,6 +56,18 @@ type Interface interface {
 	GetPackage(*fastly.GetPackageInput) (*fastly.Package, error)
 	UpdatePackage(*fastly.UpdatePackageInput) (*fastly.Package, error)
 
+	CreateDictionary(*fastly.CreateDictionaryInput) (*fastly.Dictionary, error)
```
These are done.

phamann commented on pkg/edgedictionary/update.go:
```diff
@@ -0,0 +1,56 @@
+	manifest manifest.Data
+	Input    fastly.UpdateDictionaryInput
+}
+
+// NewUpdateCommand returns a usable command registered under the parent.
+func NewUpdateCommand(parent common.Registerer, globals *config.Data) *UpdateCommand {
+	var c UpdateCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("update", "Update name of dictionary on a Fastly service version").Alias("get")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("version", "Number of service version").Required().IntVar(&c.Input.Version)
+	c.CmdClause.Flag("name", "Old name of Dictionary").Short('n').Required().StringVar(&c.Input.Name)
+	c.CmdClause.Flag("new-name", "New name of Dictionary").Required().StringVar(&c.Input.NewName)
+	return &c
+}
+
+// Exec invokes the application logic for the command.
+func (c *UpdateCommand) Exec(in io.Reader, out io.Writer) error {
+	serviceID, source := c.manifest.ServiceID()
+	if source == manifest.SourceUndefined {
+		return errors.ErrNoServiceID
+	}
+	c.Input.Service = serviceID
+
+	d, err := c.Globals.Client.UpdateDictionary(&c.Input)
+	if err != nil {
+		return err
+	}
+
+	text.Success(out, "Updated dictionary %s to %s (service %s version %d)", c.Input.Name, d.Name, d.ServiceID, d.Version)
+
+	if c.Globals.Verbose() {
+		text.Output(out, "Service ID: %s", d.ServiceID)
+		text.Output(out, "Version: %d", d.Version)
+		text.PrintDictionary(out, "", d)
+	}
+
+	return nil
+}
```
This logic will need to change now that we've upgraded to `go-fastly/v2` in #169. Specifically, all update structs now use pointers in their field values.  If you look at a foo/update.go file in that PR you will see the new2 pattern which uses our optional types.

benzvan replied:
I'm noticing these don't use kingpin's `Required()` and instead relies on its own `WasSet` to fail if the value isn't set. Is that for a future migration off of kingpin? Otherwise, it seems like unnecessary code.

Integralist replied:
@benzvan 

There was a change in Go-Fastly 2.0.0 where [basic type](https://tour.golang.org/basics/11) fields that are _optional_ on 'update' input structs, were changed to pointers using some new [helper functions](./fastly/basictypes_helper.go). 

This was done to avoid unexpected behaviours when dealing with their zero value (see [this reference](https://willnorris.com/2014/05/go-rest-apis-and-pointers/) for more details).

So here's where things get a bit complicated (thanks Kingpin), and I might get some of the details wrong as I'm not overly familiar with Kingpin at this point myself. So @phamann feel free to chime in and correct me at any point.

Because of that change in Go-Fastly, we can no longer have Kingpin assign a flag value directly to the relevant Go-Fastly struct field because it's unable to write to a pointer. This  means these Go-Fastly optional pointer fields now need to be populated indirectly by CLI logic.

To do that we define the optional fields on command struct using the `common.OptionalString` type (or [one of the other 'Optional<T>' types](https://github.com/fastly/cli/blob/782b9ed756761555fe6c45c6cc9c2c1480ca64fc/pkg/common/command.go#L71-L105) we have).

We now assign the CLI flag value to the `Value` field on the `common.OptionalString` type and we make Kingpin execute a 'callback' function using its `Action()` method to tell us the flag was provided (as I don't believe Kingpin natively gives us a way to tell if a flag was specified by a user).

Example:

```go
type UpdateCommand struct {
  ...

  Comment common.OptionalString
}
```

```diff
- c.CmdClause.Flag("comment", "A descriptive note").StringVar(&c.updateInput.Comment)
+ c.CmdClause.Flag("comment", "A descriptive note").Action(c.Comment.Set).StringVar(&c.Comment.Value)
```

You'll see the callback we execute (`c.Comment.Set`) in this case references a `common.Optional` struct method called `Set` that is embedded into the `common.OptionalString` type.

What that does is store off whether the flag value was actually set using a `WasSet` field. 

Later on in the relevant receiver's `Exec` method we can now set the Go-Fastly struct field using our type-to-pointer helper functions.

Example:

```diff
- if c.updateInput.Comment == "" { c.updateInput.Comment = d.Comment }
+ if c.Comment.WasSet { c.updateInput.Comment = fastly.String(c.Comment.Value) }
```

benzvan replied:
Thank you for that explanation of these changes. I'm still unclear about the `Required` function. In my updated code, I have this:

```
c.CmdClause.Flag("new-name", "New name of Dictionary").Required().Action(c.newname.Set).StringVar(&c.newname.Value)
```
So I'm still using `Flag.Required()` to make sure the `--new-name` flag was passed but then also using the `Action` function to call `c.newname.Set`.

In my tests, I verify that it fails if `--new-name` is not passed but I don't use `c.newname.WasSet` because if it gets past the command line parsing, it should be set.

Maybe this is partially a stylistic change and I should drop the `Required()` from that line and add the `WasSet` below?

phamann commented on pkg/edgedictionaryitem/update.go:
```diff
@@ -0,0 +1,50 @@
+package edgedictionaryitem
```
As per my other comment, this file may need to be refactored to use the the pointer pattern.

phamann reviewed (changes requested):
Hey @benzvan. As always, apologies for the delay here, @Integralist has now gone on paternity leave (👶 🎉 ) so I'm picking this back up for you.  

You may have noticed #169 which has migrated the CLI codebase to use v2 of the go-fastly client. It shouldn't be too much work, but you'll need to update this PR and specifically the import paths to use v2, the other major change is the use of pointers in the update struct field values. I've left some comments inline to discuss this.

I'm very conscious of how long this PR has taken and don't think it's fair that we keep on moving the goal posts, so please let me know if you'd prefer us to do the v2 refactoring for you.

Thank you for your work and patience :)

benzvan commented:
Just FYI: The example request json for the bulk update dictionary item endpoint is invalid in the docs. It has a trailing comma.
https://developer.fastly.com/reference/api/dictionaries/dictionary-item/#bulk-update-dictionary-item

benzvan commented:
This is ready for another review.

benzvan commented:
Just thought I'd bump this @phamann @Integralist

Integralist commented:
👋🏻 @benzvan 

I'm now _back_ from paternity leave! 🥳  

And yes, I've got this PR scheduled in my calendar to be re-reviewed and for which I aim to start tomorrow morning (UK time), so hopefully we can get this over the finish line very soon! 🙂

benzvan commented:
Congrats on your new parentage @Integralist !

Thanks for the note. I just wanted to make sure it hadn't been forgotten. :-D

Integralist commented:
Just a heads up that it seems the Windows test run has failed because of a difference in error message...

```
want "open missingFile: no such file or directory",
have "open missingFile: The system cannot find the file specified."
``` 

Which is interesting, as it's an OS specific error 🤔 

Here is the full error:

```
--- FAIL: TestDictionaryItemBatchModify (0.08s)
    --- FAIL: TestDictionaryItemBatchModify/dictionaryitem_batchmodify_--service-id_123_--dictionary-id_456_--file_missingFile (0.01s)
        edgedictionaryitem_test.go:299: want "open missingFile: no such file or directory", have "open missingFile: The system cannot find the file specified."
```

Integralist commented:
re:

> The example request json for the bulk update dictionary item endpoint is invalid.

Thanks for notifying me of this! I've got a PR to correct that now 👍🏻

benzvan commented:
That's really odd...I guess go is passing the system error back up as an error, which is going to be hard to test given the system error is different depending on the system. Maybe I can catch the error type and replace it with a consistent message.

Integralist commented:
@benzvan didn't want you to miss this comment: https://github.com/fastly/cli/pull/159#discussion_r539378242

Integralist commented:
@benzvan re:

> So I'm still using Flag.Required() to make sure the --new-name flag was passed but then also using the Action function to call c.newname.Set.

So this might just be a simple misunderstanding regarding the Dictionary update command. Let me try and explain... 

If I look at the Go-Fastly code for updating a dictionary I see the following...

```go
// UpdateDictionaryInput is used as input to the UpdateDictionary function.
type UpdateDictionaryInput struct {
	// ServiceID is the ID of the service (required).
	ServiceID string

	// ServiceVersion is the specific configuration version (required).
	ServiceVersion int

	// Name is the name of the dictionary to update.
	Name string

	NewName   *string      `form:"name,omitempty"`
	WriteOnly *Compatibool `form:"write_only,omitempty"`
}
```

This tells me that _both_ the 'new name' and the 'write only' fields are optional and thus using `Required()` in the CLI for the `--new-name` flag isn't actually necessary in this case, because strictly speaking they should be able to specify a `--write-only` flag by itself (or both `--write-only` and `--new-name` together).

Integralist commented on pkg/edgedictionary/create.go:
```diff
@@ -0,0 +1,48 @@
+	c.CmdClause = parent.Command("create", "Create a Fastly edge dictionary on a Fastly service version")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("version", "Number of service version").Required().IntVar(&c.Input.ServiceVersion)
+	c.CmdClause.Flag("name", "Name of Dictionary").Short('n').Required().StringVar(&c.Input.Name)
```
I think we need an extra flag here for the `WriteOnly` field (it's not mandatory as far as I can tell from the API documentation).

benzvan replied:
Added

Integralist commented on pkg/edgedictionary/describe.go:
```diff
@@ -0,0 +1,78 @@
+		infoInput := fastly.GetDictionaryInfoInput{
+			ServiceID:      c.Input.ServiceID,
+			ServiceVersion: c.Input.ServiceVersion,
+			ID:             c.manifest.Flag.ServiceID,
```
I don't think this should be ServiceID. Looking at the Go-Fastly implementation of the `GetDictionaryInfoInput` it states:

> ID is the alphanumeric string identifying a dictionary.

benzvan replied:
Good catch! I did not see that error. Things would have worked quite poorly this way.

benzvan replied:
Fixed with additional validation in the test

benzvan commented:
> So this might just be a simple misunderstanding regarding the Dictionary update command. Let me try and explain...

OK, this makes sense. I didn't implement the `--write-only` option in the CLI, so `--new-name` is required for it to do anything.

Integralist commented on pkg/edgedictionary/edgedictionary_test.go:
```diff
@@ -0,0 +1,444 @@
+}
+
+// These two are responses from other packages that will need to update if
+// the responses ever change.
```
I'm not sure I quite understood this comment, could you possible rephrase it or clarify for me please.

benzvan replied:
I had to re-read it myself, so I'll clarify the comment and split it so it covers each function. I was attempting to point out that these methods mock fastly api calls that are not part of the local edgedictionary package.

benzvan replied:
Updated. Please let me know if the comments are unclear or unnecessary.

Integralist commented on pkg/edgedictionary/update.go:
```diff
@@ -0,0 +1,60 @@
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("version", "Number of service version").Required().IntVar(&c.input.ServiceVersion)
+	c.CmdClause.Flag("name", "Old name of Dictionary").Short('n').Required().StringVar(&c.input.Name)
+	c.CmdClause.Flag("new-name", "New name of Dictionary").Required().Action(c.newname.Set).StringVar(&c.newname.Value)
```
As explained in a separate conversation thread, `--new-name` is optional along with another flag not yet defined (but should be added here) `--write-only` and because there are two optional flags it means `--new-name` no longer is 'required' for this command to be useful.

```suggestion
	c.CmdClause.Flag("new-name", "New name of Dictionary").Action(c.newname.Set).StringVar(&c.newname.Value)
```

benzvan replied:
Updated.

Integralist commented on pkg/edgedictionary/update.go:
```diff
@@ -0,0 +1,60 @@
+	}
+	c.input.ServiceID = serviceID
+
+	c.input.NewName = &c.newname.Value
```
As per my comments elsewhere let's utilize the `WasSet` code checks as found elsewhere and also add `WriteOnly` field.

Integralist commented on pkg/edgedictionaryitem/describe.go:
```diff
@@ -0,0 +1,49 @@
+	c.CmdClause = parent.Command("describe", "Show detailed information about a Fastly edge dictionary item").Alias("get")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("name", "Name of dictionary").Short('n').Required().StringVar(&c.Input.DictionaryID)
+	c.CmdClause.Flag("key", "Dictionary item key").Required().StringVar(&c.Input.ItemKey)
```
Is there any reason we don't provide a `-k` short key like we do for `--name` with `-n`? 

If there's no reason not to, then could you just go back and ensure we have `.Short()` equivalents wherever necessary for consistency.

benzvan replied:
Yes. I didn't use a `-k` for `--key` because that would suggest `-v` for `--value`, which is already taken by `--verbose` which is a global option.

Integralist commented on pkg/edgedictionaryitem/describe.go:
```diff
@@ -0,0 +1,49 @@
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("describe", "Show detailed information about a Fastly edge dictionary item").Alias("get")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("name", "Name of dictionary").Short('n').Required().StringVar(&c.Input.DictionaryID)
```
This should probably more closely align with Go-Fastly and the API in that it's actually an ID and not a name.

```suggestion
	c.CmdClause.Flag("dictionary-id", "Dictionary ID").Short('d').Required().StringVar(&c.Input.DictionaryID)
```

Integralist replied:
BTW if I missed this 'name' vs 'id' problem any where else could you just presume a similar change is necessary there too.

**UPDATE**: I've just noticed in another file for listing dictionaries we do actually use `dictionary-id` as a flag.

benzvan replied:
Yea...this was my mistake copy/pasting when some actions take a name and some take an ID.

Integralist commented on pkg/edgedictionaryitem/update.go:
```diff
@@ -0,0 +1,54 @@
+	itemvalue common.OptionalString
+}
+
+// NewUpdateCommand returns a usable command registered under the parent.
```
As per the following comment...
```suggestion
// NewUpdateCommand returns a usable command registered under the parent.
//
// TODO(integralist) update to not use common.OptionalString once we have a
// new Go-Fastly release that modifies UpdateDictionaryItemInput so that the 
// ItemValue is no longer optional.
```

Integralist commented on pkg/edgedictionaryitem/update.go:
```diff
@@ -0,0 +1,54 @@
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("dictionary-id", "Dictionary ID").Required().StringVar(&c.Input.DictionaryID)
+	c.CmdClause.Flag("key", "Dictionary item key").Required().StringVar(&c.Input.ItemKey)
+	c.CmdClause.Flag("value", "Dictionary item value").Required().Action(c.itemvalue.Set).StringVar(&c.itemvalue.Value)
```
So annoyingly the Go-Fastly implementation has the item 'value' as an optional field although I believe that to be a mistake and in fact it should be a required field (because why bother calling the api without an updated value for the key -- I don't believe it's even possible to set an empty string as a value for an edge dict key either).

For now let's keep the `.Required()` as we want the command to be useful (i.e. don't allow someone to make request without a value) but we'll still need the `common.OptionalString` for now as the version of Go-Fastly we have is internally using a pointer. When I eventually cut a new Go-Fastly release I'll come back (as per the `TODO(integralist)` comment above) and fix up this section.

Integralist commented on pkg/text/dictionaryitem_test.go:
```diff
@@ -0,0 +1,31 @@
+func TestPrintDictionaryItem(t *testing.T) {
+	for _, testcase := range []struct {
+		name           string
+		prefix         string
```
Should we delete this as we don't actually use it?

benzvan replied:
Deleted

Integralist reviewed (commented):
Thanks @benzvan for your hard work. I've left a few more comments. But it looks like we're getting very close now 🤞🏻

benzvan commented on pkg/app/run_test.go:
```diff
@@ -559,6 +561,89 @@ COMMANDS
+    -s, --service-id=SERVICE-ID  Service ID
+        --version=VERSION        Number of service version
+    -n, --name=NAME              Old name of Dictionary
+        --new-name=NEW-NAME      New name of Dictionary
```
That would be exactly why. I didn't understand what it was for and I've never used it. I'll add it in for completeness now that I get it.

benzvan replied:
Added

benzvan commented:
@Integralist I'm reasonably certain I addressed all your comments.

Integralist reviewed (commented):
@benzvan please don't hate me 🙂 but there's three _tiny_ textual tweaks I'd like to make. 

Once you've got those made I'll be able to merge and then cut a new release 🎉

benzvan commented on pkg/edgedictionaryitem/create.go:
```diff
@@ -0,0 +1,50 @@
+	"github.com/fastly/go-fastly/v2/fastly"
+)
+
+// CreateCommand calls the Fastly API to create a dictionary.
```
I can totally believe I overlooked this.

benzvan commented:
@Integralist Not at all. Except for the remediation message, which I wasn't sure how to word, the other two would have bugged me once I noticed them.

Integralist reviewed (approved):
@benzvan I think we've made it over the finish line 🎉 

Thank you for your patience and hard work on this PR Ben ❤️

Integralist commented:
@benzvan v0.21.0 CLI release can be found [here](https://github.com/fastly/cli/releases/tag/v0.21.0).
